Debulleting Porch Architecture & Components - Porch API server (#1041)
nephio-project/porch: docs/content/en/docs/5_architecture_and_components/porch-apiserver/functionality.md

- **Update**: Validates resource version, calls Engine to update package revision, returns updated resource
- **Delete**: Calls Engine to delete package revision, returns delete status

> Watch support delegates to the Engine cache for package revision watches. This system filters events based on defined watch criteria, delivers real-time change notifications, and automatically cleans up resources upon client disconnection.

Reviewer: ```suggestion
The Porch API server supports watches by delegating to the Engine cache. This system filters events based on defined watch criteria, delivers real-time change notifications, and automatically cleans up resources upon client disconnection.
```

Author: updated

> Regarding storage characteristics, there is no direct etcd storage; instead, it delegates to the Engine. The Engine is responsible for managing package data in Git via its Cache, supports all standard Kubernetes operations, and fully implements the storage.Interface.

Reviewer: ```suggestion
The storage of PackageRevisions is also delegated to the Engine; there is no direct etcd storage. The Engine is responsible for managing package data in Git via its Cache, supports all standard Kubernetes operations, and fully implements the storage.Interface.
```

Author: updated

- **List**: Lists resources for package revisions
- **Update**: Updates package content via Engine

> Content handling involves storing resources as a map where each filename is associated with its content. This design supports large package content, bypassing limitations that might be imposed by etcd. Furthermore, any updates to this content automatically trigger the execution of the render pipeline, and the system provides a RenderStatus along with the function results.

Reviewer: ```suggestion
Package contents are handled by storing resources as a map where each filename is associated with its content. This design supports large package content, bypassing limitations that might be imposed by etcd. Furthermore, any updates to this content automatically trigger the execution of the render pipeline, and the system provides a RenderStatus along with the function results.
```

Author: updated

> Regarding storage characteristics, the system is primarily read-only for most operations. Updates are exclusively permitted on Draft packages. Content is retrieved on-demand, meaning it is not cached within the API server, and all storage-related operations are delegated to the Engine.

Reviewer: ```suggestion
PackageRevisionResources storage is handled such that operations are primarily read-only. Updates are exclusively permitted on Draft packages. Content is retrieved on-demand, meaning it is not cached within the API server, and all storage-related operations are delegated to the Engine.
```

Author: updated

### Create Validation

> **Validation rules:** Include requiring a resource version for optimistic locking, ensuring lifecycle transition validity, enforcing immutability constraints for Published packages, making tasks append-only (meaning tasks cannot be removed), and restricting metadata updates.

Author: corrected

> **Validation rules:** Include ensuring required fields are present (such as package name and repository), adhering to lifecycle constraints (preventing creation of Published/DeletionProposed states), validating tasks (allowing a maximum of one task), confirming the correct workspace name format, and verifying package path validity.

Reviewer: ```suggestion
**Validation rules** include ensuring required fields are present (such as package name and repository), adhering to lifecycle constraints (preventing creation of Published/DeletionProposed states), validating tasks (allowing a maximum of one task), confirming the correct workspace name format, and verifying package path validity.
```

Author: updated the way we discussed it on teams

> **Validation process:** Calling strategy is validated before any Engine operation. Specific field errors are returned for invalid specifications, effectively preventing invalid resources from reaching the Engine and providing clear error messages to clients.

Reviewer: ```suggestion
**Validation process:** Validation is done for each strategy before any Engine operation. Specific field errors are returned for invalid specifications, effectively preventing invalid resources from reaching the Engine and providing clear error messages to clients.
```

Author: updated

### Update Validation

> **Validation rules:** Include requiring a resource version for optimistic locking, ensuring lifecycle transition validity, enforcing immutability constraints for Published packages, making tasks append-only (meaning tasks cannot be removed), and restricting metadata updates.

Reviewer: ```suggestion
**Validation rules** include requiring a resource version for optimistic locking, ensuring lifecycle transition validity, enforcing immutability constraints for Published packages, making tasks append-only (meaning tasks cannot be removed), and restricting metadata updates.
```

Author: updated the way we discussed it on teams

> **Validation process:** Involves invoking strategy validation prior to any Engine update operation. This process meticulously compares old and new objects to confirm that all proposed changes are permissible, and it subsequently returns field errors for any updates that are deemed invalid.

Reviewer: ```suggestion
The **validation process** involves invoking strategy validation prior to any Engine update operation. This process meticulously compares old and new objects to confirm that all proposed changes are permissible, and it subsequently returns field errors for any updates that are deemed invalid.
```

Author: left the wat it is after our conversation on teams

### Status Validation

> **Validation rules:** Status subresource updates should be validated separately. Condition format, RenderStatus structure as well as DownstreamTargets should be validated.

Reviewer: ```suggestion
**Validation rules:** Status subresource updates are validated separately. This includes condition format, RenderStatus structure as well as DownstreamTargets.
```

Author: updated

> **Validation process:** Involves calling Strategy.ValidateStatusUpdate for all status updates. This ensures that status updates do not inadvertently modify the specification, validates the overall status structure, and returns detailed field errors when an invalid status is detected.

Reviewer: ```suggestion
The **validation process** involves calling `Strategy.ValidateStatusUpdate` for all status updates. This ensures that status updates do not inadvertently modify the specification, validates the overall status structure, and returns detailed field errors when an invalid status is detected.
```

Author: left the wat it is after our conversation on teams

### PrepareForCreate

> **Operations:** Generating a name if one is not provided, setting the default lifecycle to Draft, initializing status conditions, setting the creation timestamp, and adding default labels/annotations.

Reviewer: ```suggestion
PrepareForCreate handles generating a name if one is not provided, setting the default lifecycle to Draft, initializing status conditions, setting the creation timestamp, and adding default labels/annotations.
```

Author: left the wat it is after our conversation on teams

> **Admission process:** Called before validation, it modifies the resource in-place to ensure a consistent initial state and prepares the resource for the Engine.

Reviewer: ```suggestion
The admission process is called before validation, it modifies the resource in-place to ensure a consistent initial state and prepares the resource for the Engine.
```

Author: left the wat it is after our conversation on teams

### PrepareForUpdate

> **Operations:** Validating the resource version, enforcing immutability rules, preserving status on specification updates, updating the modification timestamp, and merging labels/annotations.

Reviewer: ```suggestion
PrepareForUpdate is responsible for validating the resource version, enforcing immutability rules, preserving status on specification updates, updating the modification timestamp, and merging labels/annotations.
```

Author: left the wat it is after our conversation on teams

> **Admission process:** Called before validation, it checks the current versus desired state, enforces business rules, and prepares the resource for the Engine.

Reviewer: ```suggestion
The admission process called before validation here as well, it checks the current versus desired state, enforces business rules, and prepares the resource for the Engine.
```

Author: left the wat it is after our conversation on teams

### Canonicalization

> **Operations:** Normalizing resource representation, removing redundant fields, appling default values, and ensuring a consistent format.

Author: corrected

> **Operations:** Normalizing resource representation, removing redundant fields, applying default values, and ensuring a consistent format.

Reviewer: ```suggestion
Canonicalization does the normalizing of the resource representation, removes redundant fields, applies default values, and ensures a consistent format.
```

Author: left the wat it is after our conversation on teams

> **Canonicalization process:** Called after validation, it ensures a consistent storage format, simplifies comparison operations, and improves cache efficiency.

Reviewer: ```suggestion
The canonicalization process is also called only after validation, it ensures a consistent storage format, simplifies comparison operations, and improves cache efficiency.
```

Author: left the wat it is after our conversation on teams

### Column Definitions

**PackageRevision columns**
| Column Name | Description           |
| :---------- | :-------------------- |
| Name        | Resource name         |
| Package     | Package name          |
| WorkspaceName | Workspace identifier  |
| Revision    | Revision number       |
| Lifecycle   | Current lifecycle state |

> | Repository  | Source repository     |

Author: table corrected

> The table format adheres to Kubernetes conventions, supports sorting and filtering, and provides human-readable output consistent with kubectl expectations.

Reviewer: ```suggestion
The table format adheres to Kubernetes conventions, supports sorting and filtering, and provides human-readable output consistent with `kubectl` expectations.
```

Author: updated

### Conversion Process

> **Conversion flow:** When kubectl requests a table format, the REST storage calls Strategy.ConvertToTable. This strategy then extracts the necessary column values and returns a metav1.Table with rows, which kubectl subsequently formats for display.

Reviewer: ```suggestion
**Conversion flow:** When `kubectl` requests a table format, the REST storage calls `Strategy.ConvertToTable`. This strategy then extracts the necessary column values and returns a `metav1.Table` with rows, which kubectl subsequently formats for display.
```

Author: updated

> **Conversion characteristics:** This conversion process supports both list and individual resources, gracefully handles missing fields, provides consistent formatting, and enables various kubectl get commands.

Reviewer: ```suggestion
**Conversion characteristics:** This conversion process supports both list and individual resources, gracefully handles missing fields, provides consistent formatting, and enables various `kubectl` get commands.
```

Author: updated

### Resource Cleanup

**Cleanup operations:** These operations include removing PackageRev CRs for deleted repositories, cleaning up orphaned cache entries, removing stale watch registrations, and garbage collecting expired resources.

**Cleanup triggers:** Various events can trigger these cleanup processes, including repository deletion, cache eviction, and watch disconnection. Periodic maintenance routines also initiate cleanup to ensure system health.

**Cleanup coordination:** Cleanup is coordinated by being triggered through lifecycle events and runs asynchronously. This approach ensures consistency across the system and effectively prevents resource leaks.

## Performance Optimization

The API Server employs several optimization strategies:

### List Operation Optimization

**Optimization techniques:** Concurrent repository listing with configurable maximum concurrency, per-repository timeouts to prevent slow repositories from blocking, early termination on context cancellation, and efficient filtering at the Engine level.

> **Configuration:** MaxConcurrentLists for maximum concurrent repository operations and ListTimeoutPerRepository which sets a timeout per repository, preventing slow repositories from impacting overall performance.

Reviewer: This may be better as a list IMO

Author: reverted to a list

### Watch Errors

**Error handling:** Registration errors are returned immediately upon occurrence. Should delivery errors arise, the watch stream is closed, an error event is sent to the client, and an automatic cleanup process is initiated.

> **Error recovery:** For error recovery, clients have the capability to re-establish the watch, resuming from the last known resource version. This mechanism ensures that there is no data loss during transient errors, allowing for graceful degradation of service.

Reviewer: ```suggestion
**Error recovery:** For error recovery, clients are able to re-establish the watch and resume from the last known resource version. This mechanism ensures that there is no data loss during transient errors, allowing for graceful degradation of service.
```

Author: updated

### Request Concurrency

**Concurrency characteristics:** Multiple clients can make requests concurrently, with each request processed independently. The Engine provides concurrency control, utilizing optimistic locking to prevent conflicts.

**Concurrency patterns:** Regarding concurrency patterns, read operations are fully concurrent, while write operations are serialized per package using an Engine mutex. Watch streams operate independently, and resource cleanup operations are also concurrent.

### Optimistic Locking

> **Locking mechanism:** When clients initiate updates, they are required to provide a resource version. The API Server then validates this version before making a call to the Engine, which in turn compares the provided version with its current version. If a mismatch is detected, a conflict is returned, necessitating the client to re-read the resource and retry the operation.

Reviewer: Maybe we could remark that this is how the base kubernetes apiserver works as well?

Author: added
